vault/acl.go

```diff
@@ -47,25 +50,58 @@ func NewACL(policies []*Policy) (*ACL, error) {
 			// Check for an existing policy
 			raw, ok := tree.Get(pc.Prefix)
 			if !ok {
-				tree.Insert(pc.Prefix, pc.CapabilitiesBitmap)
+				tree.Insert(pc.Prefix, pc.Permissions)
 				continue
 			}
-			existing := raw.(uint32)
+
+			// these are the ones already in the tree
+			existingPerms := raw.(*Permissions)
 
 			switch {
-			case existing&DenyCapabilityInt > 0:
+			case existingPerms.CapabilitiesBitmap&DenyCapabilityInt > 0:
 				// If we are explicitly denied in the existing capability set,
 				// don't save anything else
+				continue
 
-			case pc.CapabilitiesBitmap&DenyCapabilityInt > 0:
+			case pc.Permissions.CapabilitiesBitmap&DenyCapabilityInt > 0:
 				// If this new policy explicitly denies, only save the deny value
-				tree.Insert(pc.Prefix, DenyCapabilityInt)
+				pc.Permissions.CapabilitiesBitmap = DenyCapabilityInt
+				pc.Permissions.AllowedParameters = nil
+				pc.Permissions.DeniedParameters = nil
+				goto INSERT
 
 			default:
 				// Insert the capabilities in this new policy into the existing
 				// value
-				tree.Insert(pc.Prefix, existing|pc.CapabilitiesBitmap)
+				pc.Permissions.CapabilitiesBitmap = existingPerms.CapabilitiesBitmap | pc.Permissions.CapabilitiesBitmap
+			}
+
+			if len(existingPerms.AllowedParameters) > 0 {
+				if pc.Permissions.AllowedParameters == nil {
+					pc.Permissions.AllowedParameters = make(map[string][]interface{}, len(existingPerms.AllowedParameters))
+				}
+
+				// Merge the two maps, appending values on key conflict.
+				for key, value := range existingPerms.AllowedParameters {
+					pc.Permissions.AllowedParameters[key] = append(value, pc.Permissions.AllowedParameters[key]...)
+				}
 			}
+
+			if len(existingPerms.DeniedParameters) > 0 {
+				if pc.Permissions.DeniedParameters == nil {
```

Comment:
Since `existingPerms` tend to contain more elements than the `pc.Permissions`, how about we do this the other way around. Append the items from `pc.Permissions` to `existingPerms` and insert `existingPerms` back into the tree?